.emacs.d/plugin/global-6.2.6/gtags-cscope/gtags-cscope.1
.\" This file is generated automatically by convert.pl from gtags-cscope/manual.in.
.TH GTAGS-CSCOPE 1 "March 2011" "GNU Project"
.SH NAME
gtags\-cscope - interactively examine a C program
.SH SYNOPSIS
\fBgtags-cscope\fP [-bCdehLlVv][-F file ][-012345678 \fIpattern\fP][-p n]
.br
.SH DESCRIPTION
\fBgtags-cscope\fP is an interactive, screen-oriented tool that allows the user to
browse through C source files for specified elements of code.
.PP
\fBgtags-cscope\fP builds the symbol cross-reference the first time it is used on
the source files for the program being browsed. On a subsequent invocation, 
\fBgtags-cscope\fP rebuilds the cross-reference only if a source file
has changed or the list of source files is different. When the
cross-reference is rebuilt, it is updated incrementally, which makes rebuilding faster
than the initial build.
.SH OPTIONS
Some command line arguments can only occur as the only argument in
the execution of \fBgtags-cscope\fP.  They cause the program to just print out
some output and exit immediately:
.TP
\fB-h\fP
View the long usage help display.
.TP
\fB-V\fP
Print on the first line of screen the version number of gtags-cscope.
.TP
\fB--help\fP
Same as \fB-h\fP
.TP
\fB--version\fP
Same as \fB-V\fP
.PP
The following options can appear in any combination:
.TP
\fB-a\fP
Print absolute path name.
.TP
\fB-b\fP
Build the cross-reference only.
.TP
\fB-C\fP
Ignore letter case when searching.
.TP
\fB-d\fP
Do not update the cross-reference.
.TP
\fB-e\fP
Suppress the <Ctrl>-e command prompt between files.
.TP
\fB-F\fP \fIfile\fP
Read symbol reference lines from \fIfile\fP. 
(A symbol reference file is created by > and >>,
and can also be read using the < command,
described under ``Issuing Subsequent Requests'', below.)
.TP
\fB-i\fP
Ignore SIGINT signal in line-oriented mode.
.TP
\fB-L\fP
Do a single search with line-oriented output when used with the
-num pattern option.
.TP
\fB-l\fP
Line-oriented interface (see ``Line-Oriented Interface'' below).
This option implies the -d option.
.TP
\fB-[0-9]\fP \fIpattern\fP
Go to input field \fInum\fP (counting from 0) and find 
\fIpattern\fP.
.TP
\fB-p\fP \fIn\fP
Display the last \fIn\fP file path components instead of
the default (1). Use \fI0\fP not to display the file name at all.
.TP
\fB-v\fP
Be more verbose in line-oriented mode.
.SH "Requesting the initial search"
After the cross-reference is ready, \fBgtags-cscope\fP will display this menu:
.PP
.nf
Find this C symbol:
Find this function definition:
Find functions called by this function (N/A):
Find functions calling this function:
Find this text string:
Change this text string:
Find this egrep pattern:
Find this file:
Find files #including this file:
.fi
.PP
Press the <Up> or <Down> keys repeatedly to move to the desired input
field, type the text to search for, and then press the <Return> key.
.SH "Issuing subsequent requests"
If the search is successful, any of these single-character commands
can be used:
.TP
0-9a-zA-Z
Edit the file referenced by the given line number.
.TP
<Space>
Display next set of matching lines.
.TP
<Tab>
Alternate between the menu and the list of matching lines
.TP
<Up>
Move to the previous menu item (if the cursor is in the menu)
or move to the previous matching line (if the cursor is in the
matching line list.)
.TP
<Down>
Move to the next menu item (if the cursor is in the menu)
or move to the next matching line (if the cursor is in the
matching line list.)
.TP
+
Display next set of matching lines.
.TP
-
Display previous set of matching lines.
.TP
^e
Edit displayed files in order.
.TP
>
Write the displayed list of lines to a file.
.TP
>>
Append the displayed list of lines to a file.
.TP
<
Read lines from a file that is in symbol reference format
(created by > or >>), just like the -F option.
.TP
^
Filter all lines through a shell command and display the
resulting lines, replacing the lines that were already there.
.TP
|
Pipe all lines to a shell command and display them without
changing them.
.TP
^g
Read lines from the result of the execution of global(1).
.PP
At any time these single-character commands can also be used:
.TP
<Return>
Move to next input field.
.TP
^n
Move to next input field.
.TP
^p
Move to previous input field.
.TP
^y
Search with the last text typed.
.TP
^b
Move to previous input field and search pattern.
.TP
^f
Move to next input field and search pattern.
.TP
^c
Toggle ignore/use letter case when searching. (When ignoring
letter case, search for ``FILE'' will match ``File'' and ``file''.)
.TP
^r
Rebuild the cross-reference.
.TP
!
Start an interactive shell (type ^d to return to gtags-cscope).
.TP
^l
Redraw the screen.
.TP
?
Give help information about gtags-cscope commands.
.TP
^d
Exit gtags-cscope.
.PP
NOTE: If the first character of the text to be searched for matches
one of the above commands, escape it by typing a  (backslash) first.
.PP
Substituting new text for old text
.PP
After the text to be changed has been typed, gtags-cscope will prompt for
the new text, and then it will display the lines containing the old
text. Select the lines to be changed with these single-character
commands:
.TP
0-9a-zA-Z
Mark or unmark the line to be changed.
.TP
*
Mark or unmark all displayed lines to be changed.
.TP
<Space>
Display next set of lines.
.TP
+
Display next set of lines.
.TP
-
Display previous set of lines.
.TP
^a
Mark or unmark all lines to be changed.
.TP
^d
Change the marked lines and exit.
.TP
<Esc>
Exit without changing the marked lines.
.TP
!
Start an interactive shell (type ^d to return to gtags-cscope).
.TP
^l
Redraw the screen.
.TP
?
Give help information about gtags-cscope commands.
.TP
 Special keys
If your terminal has arrow keys that work in vi, you can use them
to move around the input fields. The up-arrow key is useful to move to
the previous
input field instead of using the <Tab> key repeatedly. If you have
<CLEAR>, <NEXT>, or <PREV> keys they will act as the ^l, +, and -
commands, respectively.
.SH "Line-Oriented interface"
The -l option lets you use gtags-cscope where a screen-oriented interface
would not be useful, for example, from another screen-oriented
program.
.PP
gtags-cscope will prompt with >> when it is ready for an input line starting
with the field number (counting from 0) immediately followed by the
search pattern, for example, ``lmain'' finds the definition of the
main function.
.PP
If you just want a single search, instead of the -l option use the -L
and -num pattern options, and you won't get the >> prompt.
.PP
For -l, gtags-cscope outputs the number of reference lines
cscope: 2 lines
.PP
For each reference found, gtags-cscope outputs a line consisting of the file
name, function name, line number, and line text, separated by spaces,
for example,
main.c main 161 main(argc, argv)
.PP
Note that the editor is not called to display a single reference,
unlike the screen-oriented interface.
.PP
You can use the c command to toggle ignore/use letter case when
searching. (When ignoring letter case, search for ``FILE'' will match
``File'' and ``file''.)
.PP
You can use the r command to rebuild the database.
.PP
gtags-cscope will quit when it detects end-of-file, or when the first
character of an input line is ``^d'' or ``q''.
.SH "ENVIRONMENT VARIABLES"
The following environment variables are the cscope origin.
.PP
.TP
\fBCSCOPE_EDITOR\fP
Overrides the EDITOR and VIEWER variables. Use this if you wish to use
a different editor with cscope than that specified by your
EDITOR/VIEWER variables.
.TP
\fBCSCOPE_LINEFLAG\fP
Format of the line number flag for your editor. By default, cscope
invokes your editor via the equivalent of ``editor +N file'', where
``N'' is the line number that the editor should jump to. This format
is used by both emacs and vi. If your editor needs something
different, specify it in this variable, with ``%s'' as a placeholder
for the line number.  Ex: if your editor needs to be invoked as
``editor -#103 file'' to go to line 103, set this variable to
``-#%s''.
.TP
\fBCSCOPE_LINEFLAG_AFTER_FILE\fP
Set this variable to ``yes'' if your editor needs to be invoked with
the line number option after the filename to be edited. To continue
the example from CSCOPE_LINEFLAG, above: if your editor needs to see
``editor file -#number'', set this environment variable. Users of most
standard editors (vi, emacs) do not need to set this variable.
.TP
\fBEDITOR\fP
Preferred editor, which defaults to vi.
.TP
\fBHOME\fP
Home directory, which is automatically set at login.
.TP
\fBSHELL\fP
Preferred shell, which defaults to sh.
.TP
\fBTERM\fP
Terminal type, which must be a screen terminal.
.TP
\fBTERMINFO\fP
Terminal information directory full path name. If your terminal
is not in the standard terminfo directory, see curses
and terminfo for how to make your own terminal description.
.TP
\fBTMPDIR\fP
Temporary file directory, which defaults to /tmp.
.TP
\fBVIEWER\fP
Preferred file display program (such as less), which overrides
EDITOR (see above).
.PP
The following environment variables are the GLOBAL origin.
.PP
.TP
\fBGTAGSROOT\fP
The root directory of the project.
.TP
\fBGTAGSDBPATH\fP
The directory on which tag files exist.
This value is ignored when \fBGTAGSROOT\fP is not defined.
.TP
\fBGTAGSLIBPATH\fP
If this variable is set, it is used as the path to search
for library functions. If the specified object is not
found in the project, \fBglobal\fP also search in these paths.
Since only \'GTAGS\' is targeted in the retrieval, this variable is
ignored when the \fB-r\fP or \fB-s\fP is specified.
.TP
\fBGTAGSCONF\fP
Configuration file. The default is \'$HOME/.globalrc\'.
.TP
\fBGTAGSLABEL\fP
Configuration label. The default is \fIdefault\fP.
.TP
\fBMAKEOBJDIRPREFIX\fP
If this variable is set, \'$MAKEOBJDIRPREFIX\' is used as the prefix
of BSD-style objdir. The default is \'/usr/obj\'.
.SH FILES
.TP
\'GTAGS\'
Tag file for object definitions.
.TP
\'GRTAGS\'
Tag file for object references.
.TP
\'GPATH\'
Tag file for path of source files.
.TP
\'GTAGSROOT\'
If environment variable \fBGTAGSROOT\fP is not set
and file \'GTAGSROOT\' exists in the same directory with \'GTAGS\'
then \fBglobal\fP sets \fBGTAGSROOT\fP to the contents of the file.
.TP
\'$HOME/.globalrc\', \'/etc/gtags.conf\', \'[sysconfdir]/gtags.conf\'
Configuration files.
.SH "SEE ALSO"
\fBgtags\fP(1),
\fBglobal\fP(1),
\fBhtags\fP(1).
.PP
GNU GLOBAL source code tag system
.br
(http://www.gnu.org/software/global/).
.SH BUG
The function field of the display is almost <unknown> since
GLOBAL doesn't recognize it.
.PP
``Find functions called by this function'' is not implemented.
.SH AUTHOR
Joe Steffen (original author) and others
.SH HISTORY
\fBCscope\fP was originally developed at Bell Labs in the early 1980's,
and was released as free software under the BSD license in April 2000.
\fBGtags-cscope\fP is a derivative of cscope to use GLOBAL as the back-end.
Its line-oriented interface was originally written in 2006, and was
re-implemented in 2011 using cscope itself.


.emacs.d/plugin/global-6.2.6/gtags-cscope/gtags-cscope.c
/*===========================================================================
 Copyright (c) 1998-2000, The Santa Cruz Operation 
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 *Redistributions of source code must retain the above copyright notice,
 this list of conditions and the following disclaimer.

 *Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 *Neither name of The Santa Cruz Operation nor the names of its contributors
 may be used to endorse or promote products derived from this software
 without specific prior written permission. 

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS
 IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION)
 HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 DAMAGE. 
 =========================================================================*/


/**	@file
 *	@NAME{gtags-cscope} - interactive C symbol cross-reference (@NAME{cscope})
 *
 *	main functions
 */

#include "global-cscope.h"

#include "build.h"
#include "version-cscope.h"	/* FILEVERSION and FIXVERSION */

/* for libutil */
#include "env.h"
#include "gparam.h"
#include "path.h"
#include "test.h"
#include "version.h"
/* usage */
#include "const.h"

#include <stdlib.h>	/* atoi */
#include <unistd.h>
#if defined(USE_NCURSES) && !defined(RENAMED_NCURSES)
#include <ncurses.h>
#else
#include <curses.h>
#endif
#include <sys/types.h>	/* needed by stat.h */
#include <sys/stat.h>	/* stat */
#include <signal.h>

#if defined(_WIN32) && !defined(__CYGWIN__)
#define S_IRWXG 0070
#define S_IRWXO 0007
#define mkdir(path,mode) mkdir(path)
#endif

/** @name defaults for unset environment variables */
/** @{ */
/** text editor */
#if defined(__DJGPP__) || (defined(_WIN32) && !defined(__CYGWIN__))
#define EDITOR	"tde"
#else
#define EDITOR	"vi"
#endif

/** no @NAME{\$HOME} --\> use root directory */
#define HOME	"/"
/** shell executable */
#define	SHELL	"sh"

/** default: used by @NAME{vi} and @NAME{emacs} */
#define LINEFLAG "+%s"
/** temp dir */
#define TMPDIR	"/tmp"
/** @} */

static char const rcsid[] = "$Id: gtags-cscope.c,v 1.27 2012/10/13 07:02:06 shigio Exp $";

char	*editor, *shell, *lineflag;	/**< environment variables */
char	*global_command;	/**< @FILE{global} by default */
char	*gtags_command;		/**< @FILE{gtags} by default */
char	*home;			/**< Home directory */
BOOL	lineflagafterfile;
char	*argv0;			/**< command name */
int	dispcomponents = 1;	/**< file path components to display */
#if CCS
BOOL	displayversion;		/**< display the C Compilation System version */
#endif
BOOL	editallprompt = YES;	/**< prompt between editing files */
BOOL	incurses = NO;		/**< in @NAME{curses} */
BOOL	isuptodate;		/**< consider the crossref up-to-date */
BOOL	linemode = NO;		/**< use line oriented user interface */
BOOL	verbosemode = NO;	/**< print extra information on line mode */
BOOL	absolutepath = NO;	/**< print absolute path name */
BOOL	ignoresigint = NO;	/**< ignore @NAME{SIGINT} signal */
BOOL	ogs;			/**< display @NAME{OGS} book and subsystem names */
char	*prependpath;		/**< prepend path to file names */
FILE	*refsfound;		/**< references found file */
char	temp1[PATHLEN + 1];	/**< temporary file name */
char	temp2[PATHLEN + 1];	/**< temporary file name */
char	tempdirpv[PATHLEN + 1];	/**< private temp directory */
char	tempstring[TEMPSTRING_LEN + 1]; /**< use this as a buffer, instead of @CODE{yytext}, 
				 * which had better be left alone */
char	*tmpdir;		/**< temporary directory */

static	BOOL	onesearch;		/**< one search only in line mode */
static	char	*reflines;		/**< symbol reference lines file */

/* Internal prototypes: */
static	void	longusage(void);
static	void	usage(void);
int	qflag;

#ifdef HAVE_FIXKEYPAD
void	fixkeypad();
#endif

#if defined(KEY_RESIZE) && defined(SIGWINCH)
void 
sigwinch_handler(int sig, siginfo_t *info, void *unused)
{
    (void) sig;
    (void) info;
    (void) unused;
    if(incurses == YES)
        ungetch(KEY_RESIZE);
}
#endif

int
main(int argc, char **argv)
{
    char *s;
    int c;
    pid_t pid;
    struct stat	stat_buf;
    mode_t orig_umask;
#if defined(KEY_RESIZE) && defined(SIGWINCH)
    struct sigaction winch_action;
#endif
	
    /* save the command name for messages */
    argv0 = argv[0];

    /* set the options */
    while (--argc > 0 && (*++argv)[0] == '-') {
	/* HBB 20030814: add GNU-style --help and --version options */
	if (strequal(argv[0], "--help")
	    || strequal(argv[0], "-h")) {
	    longusage();
	    myexit(0);
	}
	if (strequal(argv[0], "--version")
	    || strequal(argv[0], "-V")) {
#if CCS
	    displayversion = YES;
#else
	    fprintf(stderr, "%s: %s (based on cscope version %d%s)\n", argv0, get_version(),
		    FILEVERSION, FIXVERSION);
	    myexit(0);
#endif
	}

	for (s = argv[0] + 1; *s != '\0'; s++) {
			
	    /* look for an input field number */
	    if (isdigit((unsigned char) *s)) {
		field = *s - '0';
		if (field > 8) {
		    field = 8;
		}
		if (*++s == '\0' && --argc > 0) {
		    s = *++argv;
		}
		if (strlen(s) > PATLEN) {
		    postfatal("\
gtags-cscope: pattern too long, cannot be > %d characters\n", PATLEN);
		    /* NOTREACHED */
		}
		strcpy(Pattern, s);
		goto nextarg;
	    }
	    switch (*s) {
	    case '-':	/* end of options */
		--argc;
		++argv;
		goto lastarg;
	    case 'a':	/* absolute path name */
		absolutepath = YES;
		break;
	    case 'b':	/* only build the cross-reference */
		buildonly = YES;
		linemode  = YES;
		break;
	    case 'c':	/* ASCII characters only in crossref */
		/* N/A */
		break;
	    case 'C':	/* turn on caseless mode for symbol searches */
		caseless = YES;
		break;
	    case 'd':	/* consider crossref up-to-date */
		isuptodate = YES;
		break;
	    case 'e':	/* suppress ^E prompt between files */
		editallprompt = NO;
		break;
	    case 'i':	/* ignore SIGINT signal */
		ignoresigint = YES;
		break;
	    case 'k':	/* ignore DFLT_INCDIR */
		/* N/A */
		break;
	    case 'L':
		onesearch = YES;
		/* FALLTHROUGH */
	    case 'l':
		linemode = YES;
		break;
	    case 'v':
		verbosemode = YES;
		break;
	    case 'o':	/* display OGS book and subsystem names */
		ogs = YES;
		break;
	    case 'q':	/* quick search */
		/* N/A */
		break;
	    case 'T':	/* truncate symbols to 8 characters */
		/* N/A */
		break;
	    case 'u':	/* unconditionally build the cross-reference */
		/* N/A */
		break;
	    case 'U':	/* assume some files have changed */
		/* N/A */
		break;
	    case 'R':
		usage();
		break;
	    case 'f':	/* alternate cross-reference file */
	    case 'F':	/* symbol reference lines file */
/*	    case 'i':	/* file containing file names */
	    case 'I':	/* #include file directory */
	    case 'p':	/* file path components to display */
	    case 'P':	/* prepend path to file names */
	    case 's':	/* additional source file directory */
	    case 'S':
		c = *s;
		if (*++s == '\0' && --argc > 0) {
		    s = *++argv;
		}
		if (*s == '\0') {
		    fprintf(stderr, "%s: -%c option: missing or empty value\n", 
			    argv0, c);
		    goto usage;
		}
		switch (c) {
		case 'f':	/* alternate cross-reference file (default: cscope.out) */
		    /* N/A */
		    break;
		case 'F':	/* symbol reference lines file */
		    reflines = s;
		    break;
		case 'i':	/* file containing file names (default: cscope.files) */
		    /* N/A */
		    break;
		case 'I':	/* #include file directory */
		    /* N/A */
		    break;
		case 'p':	/* file path components to display */
		    if (*s < '0' || *s > '9' ) {
			fprintf(stderr, "\
%s: -p option: missing or invalid numeric value\n", 
				argv0);
			goto usage;
		    }
		    dispcomponents = atoi(s);
		    break;
		case 'P':	/* prepend path to file names */
		    /* N/A */
		    break;
		case 's':	/* additional source directory */
		case 'S':
		    /* N/A */
		    break;
		}
		goto nextarg;
	    default:
		fprintf(stderr, "%s: unknown option: -%c\n", argv0, 
			*s);
	    usage:
		usage();
		fprintf(stderr, "Try the -h option for more information.\n");
		myexit(1);
	    } /* switch(option letter) */
	} /* for(option) */
    nextarg:	
	;
    } /* while(argv) */

 lastarg:
    /* read the environment */
    editor = mygetenv("EDITOR", EDITOR);
    editor = mygetenv("VIEWER", editor); /* use viewer if set */
    editor = mygetenv("CSCOPE_EDITOR", editor);	/* has last word */
    home = mygetenv("HOME", HOME);
    global_command = mygetenv("GTAGSGLOBAL", "global");
    gtags_command = mygetenv("GTAGSGTAGS", "gtags");
#if defined(_WIN32) || defined(__DJGPP__)
    shell = mygetenv("COMSPEC", SHELL);
    shell = mygetenv("SHELL", shell);
    tmpdir = mygetenv("TMP", TMPDIR);
    tmpdir = mygetenv("TMPDIR", tmpdir);
#else
    shell = mygetenv("SHELL", SHELL);
    tmpdir = mygetenv("TMPDIR", TMPDIR);
#endif
    lineflag = mygetenv("CSCOPE_LINEFLAG", LINEFLAG);
    lineflagafterfile = getenv("CSCOPE_LINEFLAG_AFTER_FILE") ? 1 : 0;

    /* make sure that tmpdir exists */
    if (lstat (tmpdir, &stat_buf)) {
	fprintf (stderr, "\
cscope: Temporary directory %s does not exist or cannot be accessed\n", 
		 tmpdir);
	fprintf (stderr, "\
cscope: Please create the directory or set the environment variable\n\
cscope: TMPDIR to a valid directory\n");
	myexit(1);
    }

    /* create the temporary file names */
    orig_umask = umask(S_IRWXG|S_IRWXO);
    pid = getpid();
    snprintf(tempdirpv, sizeof(tempdirpv), "%s/cscope.%d", tmpdir, pid);
    if(mkdir(tempdirpv,S_IRWXU)) {
	fprintf(stderr, "\
cscope: Could not create private temp dir %s\n",
		tempdirpv);
	myexit(1);
    }
    umask(orig_umask);

    snprintf(temp1, sizeof(temp1), "%s/cscope.1", tempdirpv);
    snprintf(temp2, sizeof(temp2), "%s/cscope.2", tempdirpv);

    /* if running in the foreground */
    if (signal(SIGINT, SIG_IGN) != SIG_IGN && ignoresigint == NO) {
	/* cleanup on the interrupt and quit signals */
	signal(SIGINT, myexit);
#ifdef SIGQUIT
	signal(SIGQUIT, myexit);
#endif
    }
    /* cleanup on the hangup signal */
#ifdef SIGHUP
    signal(SIGHUP, myexit);
#endif

    /* ditto the TERM signal */
    signal(SIGTERM, myexit);

    if (linemode == NO) {
	signal(SIGINT, SIG_IGN);	/* ignore interrupts */
#ifdef SIGPIPE
	signal(SIGPIPE, SIG_IGN);/* | command can cause pipe signal */
#endif

#if defined(KEY_RESIZE) && defined(SIGWINCH)
	winch_action.sa_sigaction = sigwinch_handler;
	sigemptyset(&winch_action.sa_mask);
	winch_action.sa_flags = SA_SIGINFO;
	sigaction(SIGWINCH,&winch_action,NULL);
#endif

	/* initialize the curses display package */
	initscr();	/* initialize the screen */
	entercurses();
#if TERMINFO
	keypad(stdscr, TRUE);	/* enable the keypad */
# ifdef HAVE_FIXKEYPAD
	fixkeypad();	/* fix for getch() intermittently returning garbage */
# endif
#endif /* TERMINFO */
#if UNIXPC
	standend();	/* turn off reverse video */
#endif
	dispinit();	/* initialize display parameters */
	setfield();	/* set the initial cursor position */
	clearmsg();	/* clear any build progress message */
	display();	/* display the version number and input fields */
    }

    /* if the cross-reference is to be considered up-to-date */
    if (isuptodate == YES) {
	char com[80];
	snprintf(com, sizeof(com), "%s -p >" NULL_DEVICE, global_command);
	if (system(com) != 0) {
	    postfatal("gtags-cscope: GTAGS not found. Please invoke again without -d option.\n");
            /* NOTREACHED */
	}
    } else {
	char buf[MAXPATHLEN];

	if (linemode == NO || verbosemode == YES)    /* display if verbose as well */
	    postmsg("Building cross-reference...");                 
	rebuild();
	if (linemode == NO )
            clearmsg(); /* clear any build progress message */
	if (buildonly == YES) {
	    myexit(0);
	}
	set_env("GTAGSROOT", getcwd(buf, sizeof(buf)));
	set_env("GTAGSDBPATH", getcwd(buf, sizeof(buf)));
    }

    /* opendatabase(); */

    /* if using the line oriented user interface so cscope can be a 
       subprocess to emacs or samuel */
    if (linemode == YES) {
	if (*Pattern != '\0') {		/* do any optional search */
	    if (search() == YES) {
		/* print the total number of lines in
		 * verbose mode */
		if (verbosemode == YES)
		    printf("cscope: %d lines\n",
			   totallines);

		while ((c = getc(refsfound)) != EOF)
		    putchar(c);
	    }
	}
	if (onesearch == YES)
	    myexit(0);
		
	for (;;) {
	    char buf[PATLEN + 2];
			
	    printf(">> ");
	    fflush(stdout);
	    if (fgets(buf, sizeof(buf), stdin) == NULL) {
		myexit(0);
	    }
	    /* remove any trailing newline character */
	    if (*(s = buf + strlen(buf) - 1) == '\n') {
		*s = '\0';
	    }
	    switch (*buf) {
	    case '0':
	    case '1':
	    case '2':
	    case '3':
	    case '4':
	    case '5':
	    case '6':
	    case '7':
	    case '8':
		field = *buf - '0';
		strcpy(Pattern, buf + 1);
		search();
		printf("cscope: %d lines\n", totallines);
		while ((c = getc(refsfound)) != EOF) {
		    putchar(c);
		}
		break;

	    case 'c':	/* toggle caseless mode */
	    case ctrl('C'):
		if (caseless == NO) {
		    caseless = YES;
		} else {
		    caseless = NO;
		}
		break;

	    case 'r':	/* rebuild database cscope style */
	    case ctrl('R'):
		rebuild();
		putchar('\n');
		break;

	    case 'C':	/* clear file names */
		/* N/A */
		putchar('\n');
		break;

	    case 'F':	/* add a file name */
		/* N/A */
		putchar('\n');
		break;

	    case 'q':	/* quit */
	    case ctrl('D'):
	    case ctrl('Z'):
		myexit(0);

	    default:
		fprintf(stderr, "gtags-cscope: unknown command '%s'\n", buf);
		break;
	    }
	}
	/* NOTREACHED */
    }
    /* do any optional search */
    if (*Pattern != '\0') {
	atfield();		/* move to the input field */
	command(ctrl('Y'));	/* search */
    } else if (reflines != NULL) {
	/* read any symbol reference lines file */
	readrefs(reflines);
    }
    display();		/* update the display */

    for (;;) {
	if (!selecting)
	    atfield();	/* move to the input field */

	/* exit if the quit command is entered */
	if ((c = mygetch()) == EOF || c == ctrl('D') || c == ctrl('Z')) {
	    break;
	}
	/* execute the commmand, updating the display if necessary */
	if (command(c) == YES) {
	    display();
	}

	if (selecting) {
	    move(displine[curdispline], 0);
	    refresh();
	}
    }
    /* cleanup and exit */
    myexit(0);
    /* NOTREACHED */
    return 0;		/* avoid warning... */
}

void
cannotopen(char *file)
{
    posterr("Cannot open file %s", file);
}

/* FIXME MTE - should use postfatal here */
void
cannotwrite(char *file)
{
    char	msg[MSGLEN + 1];

    snprintf(msg, sizeof(msg), "Removed file %s because write failed", file);
    myperror(msg);	/* display the reason */

    unlink(file);
    myexit(1);	/* calls exit(2), which closes files */
}

/** enter curses mode */
void
entercurses(void)
{
    incurses = YES;
#ifndef __MSDOS__ /* HBB 20010313 */
    nonl();		    /* don't translate an output \n to \n\r */
#endif
    raw();			/* single character input */
    noecho();			/* don't echo input characters */
    clear();			/* clear the screen */
    mouseinit();		/* initialize any mouse interface */
    drawscrollbar(topline, nextline);
}


/** exit curses mode */
void
exitcurses(void)
{
	/* clear the bottom line */
	move(LINES - 1, 0);
	clrtoeol();
	refresh();

	/* exit curses and restore the terminal modes */
	endwin();
	incurses = NO;

	/* restore the mouse */
	mousecleanup();
	fflush(stdout);
}


/** normal usage message */
static void
usage(void)
{
	fputs(usage_const, stderr);
}


/** long usage message */
static void
longusage(void)
{
	fputs(usage_const, stdout);
        fputs(help_const, stdout);
}

/** cleanup and exit */

void
myexit(int sig)
{
	/* HBB 20010313; close file before unlinking it. Unix may not care
	 * about that, but DOS absolutely needs it */
	if (refsfound != NULL)
		fclose(refsfound);
	
	/* remove any temporary files */
	if (temp1[0] != '\0') {
		unlink(temp1);
		unlink(temp2);
		rmdir(tempdirpv);
	}
	/* restore the terminal to its original mode */
	if (incurses == YES) {
		exitcurses();
	}
	/* dump core for debugging on the quit signal */
#ifdef SIGQUIT
	if (sig == SIGQUIT) {
		abort();
	}
#endif
	/* HBB 20000421: be nice: free allocated data */
	exit(sig);
}
